GitUI/FilterBranchHelper.cs

```diff
@@ -164,10 +166,10 @@ private void ApplyBranchFilter(bool refresh)
             }
         }
 
-        private void UpdateBranchFilterItems()
+        private async Task UpdateBranchFilterItemsAsync()
         {
             string filter = _NO_TRANSLATE_toolStripBranches.Items.Count > 0 ? _NO_TRANSLATE_toolStripBranches.Text : string.Empty;
-            var branches = GetBranchAndTagRefs(localToolStripMenuItem.Checked, tagsToolStripMenuItem.Checked, remoteToolStripMenuItem.Checked);
+            var branches = await GetBranchAndTagRefs(localToolStripMenuItem.Checked, tagsToolStripMenuItem.Checked, remoteToolStripMenuItem.Checked);
```

Comment:
We have a class [AsyncLoader](https://github.com/gitextensions/gitextensions/blob/master/GitCommands/AsyncLoader.cs#L7) that helps us dealing with cancellation and disposed controls. It cancels the previous task if a new is scheduled. It implements IDisposable, so compiler warns if it is not disposed. It cancels the current task on being disposed. The `async await` does not provide this functionality automagically. In places like this we have to at least take care of the Form(or Control) being disposed.
That means we can't execute the continuation code not ensuring that the Form has not been disposed.

Reply:
I took a look at `AsyncLoader`. It does support cancellation, though only at the top level, and only if you use a long-lived instance of the type. The static `AsyncLoader.DoAsync` methods don't allow for any of the cancellation behaviour. I have only replaced such fire-and-forget occurrences in this PR, so the behaviour should be identical.

Reply:
My main concern here is that there is no checking for the form not being disposed. The lack of cancellation is a minor issue. It was quite prevalent to get an ObjectDisposedException. That cases was handled either by disposing AsyncLoader or using InvokeAsync.

Reply:
I see. Yes, that's true of changes that unblock the UI thread via tasks. The continuation should always assume that the situation may have changed (controls disposed, results out of order, etc) and take defensive steps. I'll look at these examples again. This extra effort is the price for keeping a responsive UI.